compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -122,28 +123,34 @@ struct LinkTuningSpecsPass final
   }
 
   void runOnOperation() override {
-    ModuleOp module = getOperation();
-    SmallVector<NamedSequenceOp> tuningSpecs;
-
-    for (ModuleOp nested : findNestedModulesWithNamedSequences(module)) {
-      llvm::append_range(tuningSpecs, findTuningSpecs(nested));
+    if (failed(linkTuningSpecs(getOperation()))) {
+      signalPassFailure();
     }
+  }
+};
 
-    for (NamedSequenceOp spec : tuningSpecs) {
-      LDBG("Found tuning spec: " << spec.getSymName());
-      if (failed(validateTuningSpec(spec))) {
-        return signalPassFailure();
-      }
-    }
+} // namespace
+
+FailureOr<NamedSequenceOp> linkTuningSpecs(ModuleOp module) {
+  SmallVector<NamedSequenceOp> tuningSpecs;
 
-    if (tuningSpecs.empty()) {
-      LDBG("No tuning specs found, exiting without linking");
-      return;
+  for (ModuleOp nested : findNestedModulesWithNamedSequences(module)) {
```

Comment:
We should document more carefully what happens when there are conflicts, i.e some ordering of the spec. From the looks of this the default gets higher precedence than the user defined. I would make it the other way round.

Reply:
I commented in the pass description that the order in the module is preserved, IE, top down:
https://github.com/iree-org/iree/blob/29229dfb84c36d90a7d60b0e03f61a7b6e0a8d58/compiler/src/iree/compiler/Codegen/Common/Passes.td#L409-L420

>  From the looks of this the default gets higher precedence than the user defined. I would make it the other way round.
Currently we don't have any default specs, so there's nothing to test. But the intention is to have default specs apply at the very end. Once we have that, I'll make sure to add a test that checks the order.